scripts/mvftoms.py

```diff
@@ -64,6 +65,48 @@ def load(dataset, indices, vis, weights, flags):
         flags[:] = dataset.flags[indices]
 
 
+@numba.jit(nopython=True, parallel=True)
+def permute_baselines(in_vis, in_weights, in_flags, cp_index, out_vis, out_weights, out_flags):
+    """Reorganise baselines and axis order.
+
+    The inputs have dimensions (time, channel, pol-baseline), and the output has shape
+    (time, baseline, channel, pol). cp_index is a 2D array which is indexed by baseline and
+    pol to get the input pol-baseline.
+
+    cp_index may contain negative indices if the data is not present, in which
+    case it is filled with 0s and flagged.
+
+    This could probably be optimised further: the current implementation isn't
+    particularly cache-friendly, and it could benefit from unrolling the loop
+    over polarisations in some way
+    """
+    # Workaround for https://github.com/numba/numba/issues/2921
+    in_flags_u8 = in_flags.view(np.uint8)
+    n_time, n_bls, n_chans, n_pols = out_vis.shape
+    bstep = 256
+    bblocks = (n_bls + bstep - 1) // bstep
+    for t in range(n_time):
+        for bblock in numba.prange(bblocks):
+            bstart = bblock * bstep
+            bstop = min(n_bls, bstart + bstep)
+            for c in range(n_chans):
+                for b in range(bstart, bstop):
+                    for p in range(out_vis.shape[3]):
```

Comment:
Would it help to swop the `for b` and `for p`? The baselines are separated into groups with the same polarisation.

Reply:
As discussed verbally: worth a try, but it looks like it's a bit slower.